File: Cloud Computing/Spark/PageRank/pagerank.scala
import org.apache.spark.sql.SparkSession
// Initialize SparkSession
val spark = SparkSession.builder.appName("PageRank").getOrCreate()

// Read input data from GCS
val inputUri = "gs://w6h2/input/input.txt"

val lines = spark.read.textFile(inputUri).rdd

// Create an RDD of (destination page, list of links) pairs
def parseLinks(link: String): (String, Array[String]) = {
    val links = link.split(" ")
    val destination = links(0)
    val source = links.slice(1, links.length)
    return(destination, source)
}

val links = lines.map(parseLinks)

// Initialize the PageRank values 1.0 for each web page
val ranks = links.mapValues(_ => 1.0)

val dampingFactor = 0.85

var currentRanks = ranks

for (i <- 0 until 2) {
  val contributions = links.join(currentRanks).flatMap { case (_, (pages, rank)) => pages.map(page => (page, rank / pages.length)) }
  val newRanks = contributions.reduceByKey(_ + _).mapValues(rank => (1 - dampingFactor) + dampingFactor * rank)
  currentRanks = currentRanks.join(newRanks).mapValues { case (oldRank, newRank) => newRank }
}

currentRanks.collect()

File: Cloud Computing/Spark/PageRank/pagerank.py
from pyspark.sql import SparkSession
import sys


# Initialize SparkSession
spark = SparkSession.builder.appName("PageRank").getOrCreate()


# Read input data from GCS
if len(sys.argv) != 2:
  raise Exception("Exactly 1 arguments are required: <inputUri>")
inputUri = sys.argv[1]


lines = spark.read.text(inputUri).rdd.map(lambda r: r[0])


# Create an RDD of (destination page, list of links) pairs
def parse_links(link):
  links = link.split(" ")
  destination = links[0]
  source = links[1:]
  return (destination, source)


links = lines.map(parse_links)


# Initialize the PageRank values 1.0 for each web page
ranks = links.map(lambda page: (page[0], 1.0))


damping_factor = 0.85


# 2 iterations
for i in range(2):
   contributions = links.join(ranks).flatMap(lambda x: [(page, x[1][1] / len(x[1][0])) for page in x[1][0]])
   ranks = contributions.reduceByKey(lambda x, y: x + y).mapValues(lambda rank: (1-damping_factor) + damping_factor * rank)
   # Print each iteration result
   print(ranks.collect())


# Print
for (page, rank) in ranks.collect():
   print("%s has rank: %s." % (page, rank))


# Stop the SparkSession
spark.stop()
